[DAL] CSV Data Source specification (#199)
oneapi-src/oneAPI-spec: source/elements/oneDAL/source/data_management/data_sources.rst

.. highlight:: cpp
.. default-domain:: cpp

.. _data-sources:

============
Data Sources
============

This section describes the types related to the :txtref:`data-source` concept.

> 

Reviewer: What's about high-level requirements for data_source implementations?

Author: I'm not sure that there are such requirements. It is very much like the algorithm descriptor - different data source will not have anything in common probably.

---------------
CSV data source
---------------

> Class ``csv::data_source`` is an api for accessing the data source represented as a :capterm:`csv file <CSV file>`.

Reviewer 2: I'd spell out api (or at least say API)

Author: Done.

CSV data source can be used in :expr:`read` operation to extract data in text format from the given input file,
process it using provided parameters (such as delimiter and parsing options), transform it into numerical representation,
and store it as an in-memory :txtref:`dataset` of a chosen type.

Supported type of in-memory dataset for :expr:`read` operation with CSV data source is :txtref:`table`.

> CSV data source requires input file name to be set in the constructor, while other parameters can be initialized with default values.

Reviewer 3: while the other parameters of the constructor such as ... rely on default values

Author: Done.

-------------
Usage example
-------------
::

   const auto data_source = onedal::csv::data_source("data.csv", ',');

   const auto table = ondedal::read<onedal::table>(data_source);

---------------------
Programming Interface
---------------------

> All type and function in this section shall be declared in the

Reviewer 2: ```suggestion
All types and functions in this section shall be declared in the
```

Author: Done.

``oneapi::dal::csv`` namespace and be available via inclusion of the
``oneapi/dal/io/csv.hpp`` header file.

::

>    enum class read_options {

Reviewer 3: do we want to explicitly specify the type used to store the data such as uint32_t or different as the base class?

Author: Done.

      default = 0,
      parse_header = 1 << 0
   };

   class data_source {
   public:
      static constexpr char default_delimiter = ',';
      static constexpr read_options default_read_options = read_options::default;

      data_source(const char *file_name,
                  char delimiter = default_delimiter,
                  read_options opts = default_read_options);

      data_source(const std::string &file_name,
                  char delimiter = default_delimiter,
                  read_options opts = default_read_options);

      std::string get_file_name() const;
      char get_delimiter() const;
      read_options get_read_options() const;
   };

.. namespace:: onedal::csv
.. class:: data_source

   .. function:: data_source(const char *file_name, char delimiter = default_delimiter, read_options opts = default_read_options)

>       Creates new instance of a csv data source with the given :expr:`file_name`, :expr:`delimiter` and read options :expr:`opts` flag.

Reviewer 2: ```suggestion
      Creates a new instance of a CSV data source with the given :expr:`file_name`, :expr:`delimiter` and read options :expr:`opts` flag.
```

Author: Done.

   .. function:: data_source(const std::string &file_name, char delimiter = default_delimiter, read_options opts = default_read_options);

>       Creates new instance of a csv data source with the given :expr:`file_name`, :expr:`delimiter` and read options :expr:`opts` flag.

Reviewer 2: ```suggestion
      Creates a new instance of a CSV data source with the given :expr:`file_name`, :expr:`delimiter` and read options :expr:`opts` flag.
```

Author: Done.

   .. member:: std::string file_name = ""

>       String which contains name of the file with the dataset for read.

Reviewer 3: to read?

Author: Done.

      Getter
         | ``std::string get_filename() const``

   .. member:: char delimiter = ','

>       Symbol that represents delimiter between continuous feature columns in the input file.

Reviewer 3: that separate features in the input file?

Author: Done.

      Getter
         | ``char get_delimter() const``

   .. member:: read_options options = read_options::default

>       Value that stores options that should be applied for input file processing. \

Reviewer 3: to be applied during reading of the input file?

Author: Done.

>       ``parse_header`` option indicates whether or not first line in the input file should be processed

Reviewer 3: option indicating whether the first line in the input file shall be processed as a header record containing the information about features

What does it mean - should not included in the dataset?

Author: Rewrote this part.

      as a header record which contains information about features and should not be included in the dataset.

      Getter
         | ``read_options get_read_options() const``

Reading :expr:`onedal::read<onedal::table>(...)`
------------------------------------------------

Result
~~~~~~

> Result of a :expr:`read` operation with a specified :code:`onedal::table` template parameter is a :txtref:`table`.

Reviewer 2: ```suggestion
The result of a :expr:`read` operation with a specified :code:`onedal::table` template parameter is a :txtref:`table`.
```

Author: Done.

Operation
~~~~~~~~~

.. namespace:: onedal
.. function:: template <typename Object, typename DataSource> \
              Object read(const DataSource& ds)

>    :tparam Object: oneDAL data object that should be produced as a result of reading from the data source.

Reviewer: The description of the section about read() says that first parameter must be a `table`, but here it is an abstract type. Probably, need to invert this logic - firstly define common interface and then say that now only `table` is supported as `Object`

Author: Done.

>    :tparam Object: oneDAL data object that should be produced as a result of reading from the data source.

Reviewer 3: use term shall when we say about reqs

Author: Done.

.. highlight:: cpp
.. default-domain:: cpp

.. _data-sources:

============
Data Sources
============

This section describes the types related to the :txtref:`data-source` concept.

Read
====

> **Read operation** is a function that transforms a data source and other arguments represented via `an args <args_>`_ object
> to `a result <result_>`_ object. An operation responsible for:

Reviewer 2: ```suggestion
**Read operation** is a function that transforms a data source and other arguments represented via `an args <args_>`_ object
to `a result <result_>`_ object. The operation is responsible for:
```

> - Execution of all data retrieval and transformation routines of the data source.
> 
> - Passing SYCL* queue to data retrieval and transformation routines.

Reviewer 2: ```suggestion
- Executing all of the data retrieval and transformation routines of the data source.

- Passing a SYCL* queue to the data retrieval and transformation routines.
```

   template <typename DatasetType, typename DataSource>
   using read_args_t = /* implementation defined */;

   template <typename DatasetType, typename DataSource>
   using read_result_t = DatasetType;

   template <typename DatasetType, typename DataSource>
   read_result_t<DatasetType, DataSource> read(
      sycl::queue& queue,
      const DataSource& data_source,
      const read_args_t<DatasetType, DataSource>& args);

   } // namespace oneapi::dal

Each operation shall satisfy the following requirements:

- An operation shall accept three parameters in the following order:

  + The SYCL* queue object.
  + The data source.
  + The `args object <args_>`_.

- An operation shall return the `result object <result_>`_.

- The ``read_args_t`` and ``read_result_t`` alias
  templates shall be used for inference of the args and return types.

------------------------
Read operation shortcuts
------------------------
In order to make the code on user side less verbose, oneDAL defines the
following overloaded functions called *shortcuts* for a read operation in addition
to the general one described in section `Read operation definition
<read_operation_definition_>`_.

> - Shortcut for execution on :term:`host <Host/Device>`. Performs the same

Reviewer 2: ```suggestion
- A shortcut for execution on :term:`host <Host/Device>`. Performs the same
```

  operation as the general function on host, but does not require passing the
  queue explicitly.

  .. code-block:: cpp

     template <typename DatasetType, typename DataSource>
     read_result_t<DatasetType, DataSource> read(
        const DataSource& data_source,
        const read_args_t<DatasetType, DataSource>& args);

> - Shortcut that allows omitting explicit args creation.

Reviewer 2: ```suggestion
- A shortcut that allows omitting explicit args creation.
```

  .. code-block:: cpp

     template <typename DatasetType, typename DataSource, typename... Args>
     read_result_t<DatasetType, DataSource> read(
        sycl::queue& queue,
        const DataSource& data_source,
        Args&&... args);

> - Shortcut that allows omitting explicit queue and args creation. This is
>   combination of two previous shortcuts.

Reviewer 2: ```suggestion
- A shortcut that allows omitting explicit queue and args creation. This is a
  combination of two previous shortcuts.
```

  .. code-block:: cpp

     template <typename DatasetType, typename DataSource, typename... Args>
     read_result_t<DatasetType, DataSource> read(
        const DataSource& data_source,
        Args&&... args);

.. _args:

----
Args
----

> - The string ``%DATA_SOURCE%`` should be substituted with the name of the data source, for example, csv. 

Reviewer 2: ```suggestion
- The string ``%DATA_SOURCE%`` should be substituted with the name of the data source, for example, ``csv``. 
```

- ``%PROPERTY_NAME%`` and ``%PROPERTY_TYPE%`` are the name and the type of one

>   of the data source args properties.

Reviewer 2: ```suggestion
  of the data source's args properties.
```

.. code-block:: cpp

   namespace oneapi::dal::%DATA_SOURCE% {

   template <typename DatasetType, typename DataSource>
   class read_args {
   public:
      read_args(
         const %PROPERTY_TYPE_1%& property_name_1,
         const %PROPERTY_TYPE_2%& property_name_2,
         /* more properties */
      )
      /* Getter & Setter for the property called `%PROPERTY_NAME_1%` */
      descriptor& set_%PROPERTY_NAME_1%(%PROPERTY_TYPE_1% value);
      %PROPERTY_TYPE_1% get_%PROPERTY_NAME_1%() const;
      /* Getter & Setter for the property called `%PROPERTY_NAME_2%` */
      descriptor& set_%PROPERTY_NAME_2%(%PROPERTY_TYPE_2% value);
      %PROPERTY_TYPE_2% get_%PROPERTY_NAME_2%() const;
      /* more properties */
   };
   } // namespace oneapi::dal::%DATA_SOURCE%

------
Result
------

The result of a :expr:`read` operation is an instance of an in-memory dataset with :code:`DatasetType` type.

Data Source Types
=================

|dal_short_name| defines a set of classes.

.. list-table::
   :header-rows: 1
   :widths: 10 70

   * - Data source type
     - Description

   * - :txtref:`CSV data source <csv-data-source>`

>      - Data source which allows reading data from text file into a :txtref:`table`.

Reviewer 2: ```suggestion
     - Data source that allows reading data from a text file into a :txtref:`table`.
```

.. highlight:: cpp
.. default-domain:: cpp

.. _data-sources:

============
Data Sources
============

This section describes the types related to the :txtref:`data-source` concept.

====
Read
====

**Read operation** is a function that transforms a data source and other arguments represented via `an args <args_>`_ object

> to `a result <result_>`_ object. An operation responsible for:

Reviewer 3: it sits in the same section as csv
In ML there is a part that describes 'anatomy' and parts discussing apis-we can do the same here - find minimum amount for this 'anatomy section' in dm from algorithm anatomy in terms of description by analogy
Re csv we wanted to align with with table section in terms of describing minimal set of requirements we are aware about, table that list possible data sources (today, it is one), etc

Author: Partially done.

     read_result_t<DatasetType, DataSource> read(
        const DataSource& data_source,
        const read_args_t<DatasetType, DataSource>& args);

- Shortcut that allows omitting explicit args creation.

  .. code-block:: cpp

     template <typename DatasetType, typename DataSource, typename... Args>
     read_result_t<DatasetType, DataSource> read(
        sycl::queue& queue,
        const DataSource& data_source,
        Args&&... args);

- Shortcut that allows omitting explicit queue and args creation. This is
  combination of two previous shortcuts.

  .. code-block:: cpp

     template <typename DatasetType, typename DataSource, typename... Args>
     read_result_t<DatasetType, DataSource> read(
        const DataSource& data_source,
        Args&&... args);

.. _args:

----
Args
----

.. code-block:: cpp

   namespace oneapi::dal::%DATA_SOURCE% {

   template <typename DatasetType, typename DataSource>
   class read_args {
   public:
      read_args(
         const %PROPERTY_TYPE_1%& property_name_1,
         const %PROPERTY_TYPE_2%& property_name_2,
         /* more properties */
      )
      /* Getter & Setter for the property called `%PROPERTY_NAME_1%` */

>       descriptor& set_%PROPERTY_NAME_1%(%PROPERTY_TYPE_1% value);

Reviewer 3: you use notation %DATA_SOURCE% and others but do not introduce it in the section. Namespaces/headers section has the example how to do that

Author: Fixed.
